packages/build-utils/src/fs/get-glob-fs.ts

```diff
@@ -0,0 +1,73 @@
+import fs from 'fs';
+import { DetectorFilesystem } from '../detectors/filesystem';
+
+type GlobFs = typeof fs;
+
+function normalizePath(path: string) {
+  // on windows, this will return a path like
+  // D:/c/package.json
+  // since we abstract the filesystem, we need to remove windows specific info from the path
+  // and let the FS decide how to process the path
+  return path.replace(/^[a-zA-Z]:/, '');
+}
+
+export function getGlobFs(_fs: DetectorFilesystem): GlobFs {
+  const readdir = (
+    path: fs.PathLike,
+    callback: (err: NodeJS.ErrnoException | null, files: string[]) => void
+  ): void => {
+    _fs
+      .readdir(normalizePath(String(path)))
+      .then(stats =>
+        callback(
+          null,
+          stats.map(stat => stat.name)
+        )
+      )
+      .catch(err => callback(err, []));
+  };
+
+  const stat = (
+    path: fs.PathLike,
+    callback: (err: NodeJS.ErrnoException | null, stats: fs.Stats) => void
+  ): void => {
+    _fs
+      .isFile(normalizePath(String(path)))
+      .then(isPathAFile => {
+        callback(null, {
+          ino: 0,
+          mode: 0,
+          nlink: 0,
+          uid: 0,
+          gid: 0,
+          rdev: 0,
+          size: 0,
+          blksize: 0,
+          blocks: 0,
+          atimeMs: 0,
+          mtimeMs: 0,
+          ctimeMs: 0,
+          birthtimeMs: 0,
+          atime: new Date(),
+          mtime: new Date(),
+          ctime: new Date(),
+          birthtime: new Date(),
+          dev: 0,
+          isBlockDevice: () => false,
+          isCharacterDevice: () => false,
+          isDirectory: () => !isPathAFile,
+          isFIFO: () => false,
+          isFile: () => isPathAFile,
+          isSocket: () => false,
+          isSymbolicLink: () => false,
+        });
+      })
+      .catch(err => callback(err, { isSymbolicLink: () => false } as fs.Stats));
+  };
+
+  return {
+    readdir: readdir as typeof fs.readdir,
+    lstat: stat as typeof fs.lstat,
+    stat: stat as typeof fs.stat,
+  } as typeof fs;
```

Comment:
`glob` only ever uses these three functions.  If we were to remove the casting, how would we define `writeFile` for example?

Something like this? 

```ts
writeFile: (...) => throw new Error("not implemented")
```

Reply:
Yeah that might be good so at least the line number will tell us what when wrong if glob ever changes its implementation. Not sure how many methods need to be added here 🤔

Reply:
I tried adding the above `writeFile: (...) => throw new Error("not implemented")` but was getting some wild errors. Going to try a couple other ways

Reply:
`type GlobFs = {
  readdir: typeof fs.readdir,
  lstat: typeof fs.lstat,
  stat: typeof fs.stat,
};`

What if we change the return type for `getGlobFs` to the above because those are the only 3 functions we care about

Reply:
Actually that wouldn't work for where we utilize `getGlobFs`
```return {
    ...fs,
    readdir: readdir as typeof fs.readdir,
    lstat: stat as typeof fs.lstat,
    stat: stat as typeof fs.stat,
  };
```

Do we not want to return any of the other properties from `fs`? There are like ~82 besides `writeFile` we would have to manually add which seems also not right to do

Reply:
> Actually that wouldn't work for where we utilize `getGlobFs`
> 
> ```
>     ...fs,
>     readdir: readdir as typeof fs.readdir,
>     lstat: stat as typeof fs.lstat,
>     stat: stat as typeof fs.stat,
>   };
> ```
> 
> Do we not want to return any of the other properties from `fs`? There are like ~82 besides `writeFile` we would have to manually add which seems also not right to do

This would definitely prevent errors, but might make `glob` return unexpected results.  The point of this function is to wrap our `DetectorFilesystem` to NodeJS's `fs`

Reply:
Lets use a Proxy here to catch unexpected method calls

Reply:
@styfle recommended we use `Proxy` instead 🙌 

```ts
return new Proxy(fs, {
    get(_target, prop) {
      switch (prop) {
        case 'readdir':
          return readdir;
        case 'lstat':
        case 'stat':
          return stat;
        default:
          throw new Error('Not Implemented');
      }
    },
  });
```